dan_layer/indexer_lib/src/transaction_autofiller.rs

```diff
@@ -0,0 +1,132 @@
+}
+
+impl<TCommitteeProvider, TVnClient> TransactionAutofiller<TCommitteeProvider, TVnClient>
+where
+    TCommitteeProvider: CommitteeProvider,
+    TVnClient: ValidatorNodeClientFactory<Addr = TCommitteeProvider::Addr>,
+    TCommitteeProvider::Addr: Display,
+{
+    pub fn new(substate_scanner: Arc<SubstateScanner<TCommitteeProvider, TVnClient>>) -> Self {
+        Self { substate_scanner }
+    }
+
+    pub async fn autofill_transaction(
+        &self,
+        original_transaction: &Transaction,
+    ) -> Result<Transaction, TransactionAutofillerError> {
+        // we will include the inputs and outputs into the "involved_objects" field
+        // note that the transaction hash will not change as the "involved_objects" is not part of the hash
+        let mut autofilled_transaction = original_transaction.clone();
+
+        // scan the network to fetch all the substates for each required input
+        // TODO: perform this loop concurrently by spawning a tokio task for each scan
+        let mut input_addresses: Vec<(SubstateAddress, u32)> = vec![];
+        let mut input_substates: Vec<Substate> = vec![];
+        for r in autofilled_transaction.meta().required_inputs() {
+            let scan_res = match r.version() {
+                Some(version) => {
+                    // if the client specifyied a version, we need to retrieve it
+                    self.substate_scanner
+                        .get_specific_substate_from_committee(r.address(), version)
+                        .await?
+                },
+                None => {
+                    // if the client didn't specify a version, we fetch the latest one
+                    self.substate_scanner.get_substate(r.address(), None).await?
+                },
+            };
+
+            if let SubstateResult::Up { substate, .. } = scan_res {
+                input_addresses.push((r.address().clone(), substate.version()));
+                input_substates.push(substate);
+            } else {
+                warn!(
+                    target: LOG_TARGET,
+                    "The substate for input requirement {} is not in UP status, skipping", r
+                );
+            }
+        }
+
+        Self::add_involved_objects(&mut autofilled_transaction, &input_addresses, SubstateChange::Exists);
+
+        // add all substates related to the inputs
+        // TODO: perform this loop concurrently by spawning a tokio task for each scan
+        // TODO: we are going to only check the first level of recursion, for composability we may want to do it
+        // recursively (with a recursion limit)
+        let mut autofilled_inputs: Vec<(SubstateAddress, u32)> = vec![];
+        let related_addresses: Vec<Vec<SubstateAddress>> = input_substates
+            .iter()
+            .map(find_related_substates)
+            .collect::<Result<_, _>>()?;
+        let related_addresses = related_addresses.into_iter().flatten();
```

Comment:
Should probably not scan for related substates if the tx creator has already provided them.

```suggestion
        let related_addresses: Vec<Vec<SubstateAddress>> = input_substates
            .iter()
            .map(find_related_substates)           
            .collect::<Result<_, _>>()?;
        let related_addresses = related_addresses.into_iter().flatten()
                    // Something like
            .filter(|s| !transactions.meta().has_substate(s))
```